File: runtime/jaroku_interceptor/pricing.py
"""Model pricing — the Python reader of the shared ``runtime/pricing.json`` table.

The Node server has its own reader (``server/src/pricing.ts``) over the SAME file, so a
per-step cost computed here and a pre-run estimate computed there can never disagree
about what a model costs. Anything that changes a price changes one file.

Three rules this module exists to enforce (doc §8: "wrong cost numbers destroy trust
instantly" — the eval comparison is where that lands hardest):

1. **An unpriced model costs UNKNOWN, not zero.** ``cost_for`` returns ``None``, which
   travels as the schema's nullable ``Step.cost``. A model whose price we don't know
   must never render as free next to one we do — that's not a rounding error, it's a
   different claim.

2. **Matching is exact-then-longest-prefix**, never unordered substring. The old table
   walked a dict and took the first ``key in model`` hit, so a future
   ``claude-haiku-4-5-fast`` could have matched whichever haiku-ish key happened to be
   iterated first. Longest prefix is deterministic and picks the most specific entry.

3. **Cached input is priced as cached input.** Anthropic bills cache reads at ~0.1x and
   cache writes at ~1.25x of the input rate; LangChain folds all three into one
   ``input_tokens`` figure. Charging the full input rate for a cache read overstates
   cost by up to 10x on a caching workload.

Loading is best-effort and never raises into the observed agent: a missing or malformed
table degrades to "every model is unpriced" (null cost), which is honest, rather than to
a guess.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path
from typing import Optional

# runtime/pricing.json — sibling of jaroku_interceptor/, like .checkpoints/ and agents/.
PRICING_PATH = Path(__file__).resolve().parent.parent / "pricing.json"

_PER_MILLION = 1_000_000.0

# USD is stored to 8 decimals throughout (0.00000001 — orders of magnitude below any real
# step cost, so per-step rounding can't accumulate into a visible aggregate error).
_USD_DP = 8


@dataclass(frozen=True)
class Price:
    """One model's rates. ``input``/``output`` are USD per MILLION tokens."""

    id: str
    provider: str
    input: float
    output: float
    cache_read_mult: float
    cache_write_mult: float
    free: bool = False


def _load() -> dict[str, Price]:
    try:
        raw = json.loads(PRICING_PATH.read_text(encoding="utf-8"))
        out: dict[str, Price] = {}
        for m in raw.get("models", []):
            try:
                out[str(m["id"])] = Price(
                    id=str(m["id"]),
                    provider=str(m.get("provider", "")),
                    input=float(m["input"]),
                    output=float(m["output"]),
                    cache_read_mult=float(m.get("cache_read_mult", 0.0)),
                    cache_write_mult=float(m.get("cache_write_mult", 1.0)),
                    free=bool(m.get("free", False)),
                )
            except (KeyError, TypeError, ValueError):
                continue  # one bad row must not void the whole table
        return out
    except Exception:
        # No table -> every model is unpriced. Honest, and never fatal to a run.
        return {}


_TABLE: dict[str, Price] = _load()


def price_for(model: str) -> Optional[Price]:
    """The most specific entry for ``model``: exact match, else longest matching prefix."""
    if not model:
        return None
    exact = _TABLE.get(model)
    if exact is not None:
        return exact
    best: Optional[Price] = None
    for key, price in _TABLE.items():
        if model.startswith(key) and (best is None or len(key) > len(best.id)):
            best = price
    return best


def is_priced(model: str) -> bool:
    """Whether a cost can be computed at all for this model."""
    return price_for(model) is not None


def cost_for(
    model: str,
    *,
    input_tokens: int,
    output_tokens: int,
    cache_read_tokens: int = 0,
    cache_write_tokens: int = 0,
) -> Optional[float]:
    """USD for one model call, or ``None`` when the model has no pricing entry.

    ``input_tokens`` is the **uncached** input only — cache reads and cache writes are
    passed separately and priced at their own multipliers. Callers that only have a
    combined figure must subtract the cached portions first (see
    ``JarokuTracer._extract_usage``), or they will overcharge cached tokens at the full
    input rate.
    """
    price = price_for(model)
    if price is None:
        return None
    usd = (
        input_tokens * price.input
        + cache_read_tokens * price.input * price.cache_read_mult
        + cache_write_tokens * price.input * price.cache_write_mult
        + output_tokens * price.output
    ) / _PER_MILLION
    return round(usd, _USD_DP)

File: server/src/pricing.ts
// Model pricing — the Node reader of the shared `runtime/pricing.json` table.
//
// The Python interceptor has its own reader (runtime/jaroku_interceptor/pricing.py) over
// the SAME file. That is the point: a per-step cost computed inside a run and a pre-run
// estimate computed here are the same arithmetic over the same numbers, so the eval
// dashboard and the "this will cost about $X" gate can never disagree.
//
// Three rules, identical to the Python side:
//   1. An unpriced model costs UNKNOWN (null), never $0. Doc §8 — "wrong cost numbers
//      destroy trust instantly"; a silent zero next to a real number is the worst case,
//      because it reads as "this provider is free" rather than "we don't know".
//   2. Matching is exact-then-longest-prefix, never unordered substring.
//   3. Cached input is priced as cached input (Anthropic: ~0.1x read, ~1.25x write).
//      Charging cache reads at the full input rate overstates cost by up to 10x.
//
// Loading is best-effort: a missing or malformed table means "every model is unpriced",
// which is honest, rather than a guess.

import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join, resolve } from "node:path";

const __dirname = dirname(fileURLToPath(import.meta.url));
/** runtime/pricing.json — the single source of truth, shared with the Python runtime. */
export const PRICING_PATH = join(resolve(__dirname, "..", ".."), "runtime", "pricing.json");

const PER_MILLION = 1_000_000;
// USD carried to 8 decimals ($0.00000001) — orders of magnitude below any real step cost,
// so per-step rounding cannot accumulate into a visible aggregate error.
const USD_DP = 8;

/** One model's rates. `input`/`output` are USD per MILLION tokens. */
export interface Price {
  id: string;
  provider: string;
  input: number;
  output: number;
  cache_read_mult: number;
  cache_write_mult: number;
  free: boolean;
}

export interface TokenBreakdown {
  /** UNCACHED input only — cache reads/writes are passed separately and priced apart. */
  inputTokens: number;
  outputTokens: number;
  cacheReadTokens?: number;
  cacheWriteTokens?: number;
}

function load(): Price[] {
  try {
    const raw = JSON.parse(readFileSync(PRICING_PATH, "utf8")) as {
      models?: Record<string, unknown>[];
    };
    const out: Price[] = [];
    for (const m of raw.models ?? []) {
      const id = typeof m.id === "string" ? m.id : null;
      const input = Number(m.input);
      const output = Number(m.output);
      // One bad row must not void the whole table.
      if (!id || !Number.isFinite(input) || !Number.isFinite(output)) continue;
      out.push({
        id,
        provider: typeof m.provider === "string" ? m.provider : "",
        input,
        output,
        cache_read_mult: Number.isFinite(Number(m.cache_read_mult)) ? Number(m.cache_read_mult) : 0,
        cache_write_mult: Number.isFinite(Number(m.cache_write_mult)) ? Number(m.cache_write_mult) : 1,
        free: m.free === true,
      });
    }
    return out;
  } catch {
    return []; // no table -> every model is unpriced. Never fatal.
  }
}

const TABLE: Price[] = load();

/** Every priced model, for UI that needs to show what is and isn't comparable. */
export function allPrices(): readonly Price[] {
  return TABLE;
}

/** The most specific entry for `model`: exact match, else longest matching prefix. */
export function priceFor(model: string): Price | null {
  if (!model) return null;
  const exact = TABLE.find((p) => p.id === model);
  if (exact) return exact;
  let best: Price | null = null;
  for (const p of TABLE) {
    if (model.startsWith(p.id) && (best === null || p.id.length > best.id.length)) best = p;
  }
  return best;
}

/** Whether a cost can be computed at all for this model. */
export function isPriced(model: string): boolean {
  return priceFor(model) !== null;
}

/** USD for one model call, or `null` when the model has no pricing entry. */
export function costFor(model: string, t: TokenBreakdown): number | null {
  const price = priceFor(model);
  if (!price) return null;
  const usd =
    (t.inputTokens * price.input +
      (t.cacheReadTokens ?? 0) * price.input * price.cache_read_mult +
      (t.cacheWriteTokens ?? 0) * price.input * price.cache_write_mult +
      t.outputTokens * price.output) /
    PER_MILLION;
  return round8(usd);
}

/** Round to the module's USD precision. Exported so aggregation rounds identically. */
export function round8(usd: number): number {
  return Number(usd.toFixed(USD_DP));
}

// ── Capability ──────────────────────────────────────────────────────────────
// What a model can DO, as opposed to what it costs. Same file, same reader, and that is a
// decision rather than convenience: the composer spec requires the reasoning budgets to live
// "in the same shared pricing/capability file as model metadata", and the reason is the reason
// the prices are here — a table that lives in TypeScript is a table the Python runtime cannot
// read, and two tables about the same models drift the first time one of them is edited.
//
// Loading is best-effort in exactly the same way. A missing `reasoning` key means "this model
// has no reasoning control", which is the safe direction: the composer hides the effort chip
// rather than sending a thinking budget to a provider that would reject it.

/** How a provider spells "think harder", or null when it does not. */
export type ReasoningKind = "thinking" | "effort";

export interface Capability {
  id: string;
  /** Null when the model exposes no reasoning control at all — §6.2 omits the chip. */
  reasoning: ReasoningKind | null;
  /** The ceiling every thinking budget is validated against before dispatch. */
  maxOutputTokens: number;
  /** For the attachment budget check. Null when nothing has recorded one. */
  contextWindow: number | null;
}

/** Jaroku's four levels, in tokens, for `thinking`-shaped providers. */
export type ReasoningBudgets = Record<string, number>;

function loadCapabilities(): { caps: Capability[]; budgets: ReasoningBudgets } {
  try {
    const raw = JSON.parse(readFileSync(PRICING_PATH, "utf8")) as {
      models?: Record<string, unknown>[];
      reasoning?: { budgets?: Record<string, unknown> };
    };
    const caps: Capability[] = [];
    for (const m of raw.models ?? []) {
      const id = typeof m.id === "string" ? m.id : null;
      if (!id) continue;
      const kind = m.reasoning;
      caps.push({
        id,
        reasoning: kind === "thinking" || kind === "effort" ? kind : null,
        // A model with no recorded ceiling gets a conservative one rather than Infinity. An
        // unbounded default would let a budget through unvalidated, which is the one thing this
        // number exists to prevent.
        maxOutputTokens: Number.isFinite(Number(m.max_output_tokens)) ? Number(m.max_output_tokens) : 4096,
        contextWindow: Number.isFinite(Number(m.context_window)) ? Number(m.context_window) : null,
      });
    }
    const budgets: ReasoningBudgets = {};
    for (const [k, v] of Object.entries(raw.reasoning?.budgets ?? {})) {
      if (Number.isFinite(Number(v))) budgets[k] = Number(v);
    }
    return { caps, budgets };
  } catch {
    return { caps: [], budgets: {} };
  }
}

const CAPS = loadCapabilities();

/**
 * The capability record for a model, matched exactly-then-longest-prefix — the SAME rule
 * `priceFor` uses, and deliberately so. A model id that prices as `claude-sonnet-5` and reasons
 * as something else would be two answers about one model.
 */
export function capabilityFor(model: string): Capability | null {
  if (!model) return null;
  const exact = CAPS.caps.find((c) => c.id === model);
  if (exact) return exact;
  let best: Capability | null = null;
  for (const c of CAPS.caps) {
    if (model.startsWith(c.id) && (best === null || c.id.length > best.id.length)) best = c;
  }
  return best;
}

/** The shared budget table. Empty when the file is missing, which reads as "no thinking". */
export function reasoningBudgets(): ReasoningBudgets {
  return CAPS.budgets;
}

/** How much context this model has, for the attachment budget check. Null when unrecorded. */
export function contextWindowFor(model: string): number | null {
  return capabilityFor(model)?.contextWindow ?? null;
}
